### PR TITLE
fix(stylelint): stylelint v14 uses incompatible cli arguments

### LINKED ISSUE #1912: scss-stylelint checker doesn't work due to --syntax option removed from stylelint v14
Thank you very much for making Flycheck :bow:

## Checklist
- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].

## Bug description
`scss-stylelint` checker with Stylelint v14 always reports `The "syntax" option is no longer available. You should install an appropriate syntax, e.g. postcss-scss, and use the "customSyntax" option`.

## Steps to reproduce
1. Use `scss-stylelint` checker with Stylelint v14

- [x] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior
`scss-stylelint` handles Stylelint correctly.

## System configuration
```
Syntax checkers for buffer ___.scss in scss-mode:

First checker to run:

  scss-stylelint
    - may enable:         yes
    - may run:            t
    - executable:         Found at /___/node_modules/.bin/stylelint
    - configuration file: Not found

Checkers that are compatible with this mode, but will not run until properly configured:

  scss-lint (automatically disabled) reset
    - may enable:         no
    - executable:         Not found
    - configuration file: Not found

  sass/scss-sass-lint (automatically disabled) reset
    - may enable:         no
    - executable:         Not found
    - configuration file: Not found

  scss (automatically disabled) reset
    - may enable: no
    - executable: Not found

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 32snapshot
Emacs version:    27.2
System:           x86_64-pc-linux-gnu
Window system:    nil

```

Emacs configuration:
- [x] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

## Additional notes
`scss-stylelint` and similar checkers seem using stylelint's `--syntax` option which is removed from Stylelint v14. (https://github.com/stylelint/stylelint/blob/main/CHANGELOG.md#1400) So do you have any plans to support Stylelint v14?

BTW I wrote a temporary checker like below for my `.emacs.d/init.el` and saw it works as I expected.

```
(flycheck-define-checker general-stylelint
  "A checker for CSS and related languages using Stylelint"
  :command ("stylelint"
            (eval flycheck-stylelint-args)
            (option-flag "--quiet" flycheck-stylelint-quiet)
            (config-file "--config" flycheck-general-stylelintrc))
  :standard-input t
  :error-parser flycheck-parse-stylelint
  :predicate flycheck-buffer-nonempty-p
  :modes (scss-mode))
(flycheck-def-config-file-var flycheck-general-stylelintrc
    (general-stylelint) nil)
(add-to-list 'flycheck-checkers 'general-stylelint)
(add-hook 'scss-mode-hook
          (lambda ()
            (flycheck-disable-checker 'scss-stylelint)))
```

### PR DESCRIPTION
**Pull request by Fuco1**
Detect the major version of stylelint, and for verion 14 don't pass
the --syntax argument.

Fixes #1912

### CODE REVIEW

**Comment by cpitclaudel** [comment]
Do we care to support versions < 14?  If so, we can merge this, but otherwise we could simply remove `--syntax` entirely?

**Comment by Fuco1** [comment]
v14 was released 8 months ago, I don't know if that's too much time or too little.  I only started using it a month ago so jumped right on v14, no idea how difficult the migration is.

**Comment by justinbarclay** [comment]
For what it's worth stylelint versions belore 14 are still getting [1 million+ downloads](https://www.npmjs.com/package/stylelint?activeTab=versions) a week on npm.

**Comment by wyuenho** [comment]
@cpitclaudel @bbatsov @marsam @jcs090218 can we get this merged please?

**Review by jcs090218 — APPROVED** [review]
LGTM! If this works, then we should merge this. However, I don't have push permission to this repo, so you would have to ask other people to review it!

One suggestion is to rebase this, so we run our latest CI.

**Comment by Fuco1** [comment]
I rebased it and pushed now.

File: flycheck.el
Comment (cpitclaudel) [inline flycheck.el:11946]: Could we use `with-temp-buffer` for this?

Comment (Fuco1) [reply flycheck.el:11946]: I did it this way because `flycheck-call-checker-process` needs to run in context of the current ("flychecked") buffer, but we want to put the output to a different buffer.  With temp buffer I can rewrite it as this

```elisp
(let ((cb (current-buffer)))
  (with-temp-buffer
    (let ((temp-buffer (current-buffer)))
      (with-current-buffer cb
        (flycheck-call-checker-process
         checker nil temp-buffer nil "--version"))
      (string-to-number (car (split-string (buffer-string) "\\."))))))
```

Maybe it's a bit cleaner?

**Comment by rekenerd** [comment]
@marsam could we please merge this PR?

**Comment by wyuenho** [comment]
@bbatsov ^

**Comment by bbatsov** [comment]
@Fuco1 I'd like to finally merge this, but so much time has passed that I'd propose to just drop support for older versions and mention this in the checker's docs.

**Comment by bbatsov** [comment]
@Fuco1 ping :-)

**Comment by Fuco1** [comment]
Sure, this is almost two years old now so the v13 style lint is almost 3 years old. Not much value in this anymore (we switched to styled components anyway :D)

**Pull request closed without merging**
